File: c/src/solutions/day1/part2.c
#include "../../../include/runner/file_utils.h"
#include "../../../include/runner/helper_utils.h"
#include <stdio.h>
#include <stdlib.h>

int compare_ints(const void *a, const void *b) {
  return (*(int *)a - *(int *)b);
}

int main() {
  char *input = read_file_content("src/solutions/day1/input");
  if (input == NULL) {
    return 1;
  }

  size_t count;
  char **lines = str_split(input, '\n', &count);
  int sum = 0;
  int *left = malloc(sizeof(int) * count);
  int *right = malloc(sizeof(int) * 100000);

  if (lines) {
    int i;
    for (i = 0; *(lines + i); i++) {
      int num1, num2;
      if (sscanf(*(lines + i), "%d   %d", &num1, &num2) == 2) {
        *(left + i) = num1;
        right[num2]++;
      } else {
        fprintf(stderr, "Failed to parse line %d: %s\n", i, *(lines + i));
        exit(1);
      }
      free(*(lines + i));
    }
    printf("\n");
    free(lines);
  }

  qsort(left, count, sizeof(int), compare_ints);
  qsort(right, count, sizeof(int), compare_ints);

  for (int i = 0; i < count; i++) {
    sum += left[i] * right[left[i]];
  }

  char buf[100];
  sprintf(buf, "%d", sum);
  write_to_file("src/solutions/day1/output-part2", buf);

  free(input);
  return 0;
}

File: go/days/day4/solutions/part2.go
package solutions

import (
	"strconv"
)

func isMasOrReverse(str ...byte) bool {
	if len(str) != 3 {
		return false
	}
	return (str[0] == 'M' && str[1] == 'A' && str[2] == 'S') || (str[0] == 'S' && str[1] == 'A' && str[2] == 'M')
}

func countMasInX(input []string, x, y int) int {
	if x-1 < 0 || x+1 >= len(input[0]) || y-1 < 0 || y+1 >= len(input) {
		return 0
	}

	// check top left to bottom right
	if isMasOrReverse(input[y-1][x-1], input[y][x], input[y+1][x+1]) &&
		// check top right to bottom left
		isMasOrReverse(input[y-1][x+1], input[y][x], input[y+1][x-1]) {
		return 1
	}

	return 0
}

func Part2(input []string) []byte {
	count := 0
	for y, line := range input {
		for x := 0; x < len(line); x++ {
			if line[x] == 'A' {
				count += countMasInX(input, x, y)
			}
		}
	}

	return []byte(strconv.Itoa(count))
}
